### DISCUSSION TITLE
using latex environment floatingfigure

### DISCUSSION THREAD

**Discussion by MichelDSchmid** [quarto-dev/quarto-cli#4854] [Q&A] [answered]
Dear all

I could figure out how to set text side by side to an image in html output. In Latex I'm doing that with \begin{floatingfigure}. 
I can use the latex-environment extension to use floatingfigure with the correct options, BUT how can I deactivate the default figure environment? fig-env="" seems not to work.
If I can't deactivating the figure environment of quarto, I get in latex:
\begin{floatingfigure}[r]{6cm}
\begin{figure}
...
\end{figure}
\end{floatingfigure}

Nesting figure environments isn't possible in latex.

Some hints how I can handle this ?

Thank you very much in advance
Mike

**Comment by cderv** [comment]
Can you share a .qmd file / project of what you are doing ? 

If I understand correctly you are using `latex-environment` extension to create a figure environment you want to replace ? 

if so, does setting `fig-env` not enough to using `floatingfigure` instead of `figure` ? Are you missing the configuration option, is that so ? `fig-pos` can be use to set the `r` value but you could also be able to use `fig-pos = "[r]{6cm}"`  for the all configuration you need. 

Anyhow, I don't think you need the latex environment extension, and definitely not to replace internal figure env handling. Maybe we are missing something and we'll add it

**Comment by MichelDSchmid** [comment]
This is what I get now:
`
\begin{floatingfigure}[[r]{6cm}]

{\centering \includegraphics[width=4cm,height=\textheight]{ch_galvanische_elemente/images/voltaicPillar.jpg}

}

\caption{Voltaische Säule}

\end{floatingfigure}
`

and that's what I wrote in the qmd-file

`![Voltaische Säule](./images/voltaicPillar.jpg){ width="4cm" fig-env="floatingfigure" fig-pos="[r]{6cm}" }`

There is now [ ] to much for latex so I can't see the figure

Currently I removed the extension again ...
Thank you very much for your help

**Comment by MichelDSchmid** [comment]
\begin{floatingfigure}**[**[r]{6cm}**]**
the first rectangle bracket is to much .... but the figure is loading now ...

**Comment by MichelDSchmid** [comment]
Is there some way to manipulate the *.tex file before running xelatex? In this case I could solve the problem with the brackets

**Comment by jdutant** [comment, marked as answer]
Looks like Quarto handles `fig-env="ENVRT" fig-pos="POS"` by producing "\begin{ENVRT}[POS]" in LaTeX output, so it can't handle the fact that `floatingfigure` has an extra argument: with `fig-pos="[r]{6cm"` you get `\begin{ENVRT}[[r]{6cm}]`. Perhaps Quarto could introduce a `fig-opt` attribute for arbitrary additions, but it's not there yet. 

A workaround would be to use [conditional content](https://quarto.org/docs/authoring/conditional.html) to provide a raw LaTeX alternative for LaTeX/PDF output.

``` markdown
::: {.content-visible unless-format="pdf"}

![Voltaische Säule](./images/voltaicPillar.jpg){ width="4cm" }

:::

::: {.content-visible when-format="pdf"}

`\begin{floatingfigure}[r]{6cm}`{=latex}
![Voltaische Säule](./images/voltaicPillar.jpg){ width="4cm" }
`\caption{Voltaische Säule}`{=latex}
`\end{floatingfigure}`{=latex}
:::
```

LaTex output: 

```latex
\begin{floatingfigure}[r]{6cm}
\includegraphics[width=4cm,height=\textheight]{./images/voltaicPillar.jpg}
\caption{Voltaische Säule}    \end{floatingfigure}
```

Important: you need **both** explicit LaTeX inlines \`...\`{=latex}, and to have no empty lines around `![title](image.jpg`. Without the first, the entire content between `\begin` and `\end` is read as raw LaTeX. That is, the following:

```markdown
\begin{floatingfigure}[r]{6cm}
![Voltaische Säule](./images/voltaicPillar.jpg){ width="4cm" }
\end{floatingfigure
```

is copied verbatime in LaTeX output:

```latex
\begin{floatingfigure}[r]{6cm}
![Voltaische Säule](./images/voltaicPillar.jpg){ width="4cm" }
\end{floatingfigure}
```

And with empty lines around images you would get an extra `\begin{figure}` environment again. This:

```markdown
`\begin{floatingfigure}[r]{6cm}`{=latex}

![Voltaische Säule](./images/voltaicPillar.jpg){ width="4cm" }

`\caption{Voltaische Säule}`{=latex}
`\end{floatingfigure}`{=latex}
```

Renders in LaTeX as:

```latex
\begin{floatingfigure}[r]{6cm}
\begin{figure}
{\centering \includegraphics[width=4cm,height=\textheight]{./images/voltaicPillar.jpg}}
\caption{Voltaische Säule}
\end{figure}
\caption{Voltaische Säule} \end{floatingfigure}
```

Which is what you got with the `latex-environment` filter, I guess.

__ALTERNATIVE__. You could use the `latex-environment` extension to do without raw LaTeX on the `when-format="pdf"`. But (a) you'd still need conditional content, for the `![title](image.jpg)` should still be converted to a figure on the HTML side, while treated as a mere image in a `{.floatingfigure}` environement, and (b) you'd still need `\caption{...}` in Raw LaTeX, and (c) you need to ensure your `![title](image.jpg)` isn't alone in a paragraph, otherwise it becomes a figure again. This would look like:

``` markdown
::: {.content-visible unless-format="pdf"}

![Voltaische Säule](./images/voltaicPillar.jpg){ width="4cm" }

:::

::: {.content-visible when-format="pdf"}

::::: {.floatingfigure options="r" arguments="6cm"}

![](./images/voltaicPillar.jpg){ width="4cm" }
\caption{Voltaische Säule}

:::::

:::
```

**Reply by cscheid** [reply]
@jdutant I like the idea of fig-opt. It wouldn't be too hard to make it work. Let me give it a shot.

**Reply by MichelDSchmid** [reply]
Thank you for the workaround. I will temporarily do it this way.
AND thank you very much for your very spontaneous attempt to optimise this.

Mike

**Reply by cderv** [reply]
Yes I think we should support the same as in the latex-environment extension. We called that arguments there https://github.com/quarto-ext/latex-environment#options-and-arguments FWIW. 

Also @cscheid I thought our Lua filter would allows defining `fig-pos` starting with square brackets for that purpose, but maybe the regex is not ok, or we do that only for layout and not figures. Need to check the code. 

Basically somehting like 
````lua
local beginEnv = "\\begin{" .. env .. "}"
  if pos then
    if not string.find(pos, "^[%[{]") then
      pos = "[" .. pos .. "]"
    end
    beginEnv = beginEnv .. pos
  end
````

Meaning
* `fig-pos` value is appended after the env definition
* If not starting with `[` or `{`, then we consider that square brackets `[ ]` needs to be added before appending. 

That would allow the flexibility without adding a new argument

**Reply by jdutant** [reply]
> @jdutant I like the idea of fig-opt. It wouldn't be too hard to make it work. Let me give it a shot.

Great. Perhaps worth having a thought about the name. The `latex-environment` extension converts `options=ONE arguments=TWO` to `[ONE]{TWO}` so perhaps having `fig-pos` for something that'll be bracketed and `fig-opt` for something that'll be neither bracketed nor braced is confusing. Better call it `fig-extra` or `fig-arg`, and put it in braces? Or have `fig-pos`, `fig-opt`, `fig-arg` and render them has `\begin{<fig-env>}[<fig-pos>,<fig-opt>]{<fig-arg>}`? (Not sure if any figure environment takes something else than pos in the optional argument.)

**Reply by cscheid** [reply]
I think this is actually just a bug in quarto. Pull request incoming, hold on :)

**Reply by cscheid** [reply]
I think it's a typo on the regex on [this line](https://github.com/quarto-dev/quarto-cli/blame/main/src/resources/filters/layout/latex.lua#L318) it hasn't been touched for three years, I bet this is it.

**Reply by cscheid** [reply]
https://github.com/quarto-dev/quarto-cli/pull/4856

**Reply by jdutant** [reply]
Indeed in `string.find(pos, "^%[{")` there the intended pattern was probably @cderv's `^[%[{]`!

**Reply by cscheid** [reply]
Hm. Is `{` a valid character for a position specifier to start with? In that case, @cderv  is correct, and mine isn't. I thought it always needed `[` and could be followed by something else. I'll use his, he's smarter than me by a mile :)

**Reply by cderv** [reply]
> Is { a valid character for a position specifier to start with? 

Yes in theory you can have environment that only have options specified with `{..}` so `\begin{env}{opt}`

Example can be found in this list (https://latex-programming.fandom.com/wiki/List_of_LaTeX_environments#Figure_environment)  by searching for `}{` in the page

Example: `array` environment

https://ecampusontario.pressbooks.pub/math1020/chapter/array-environment/

> Arrays line items up in columns. Here are some basic steps for making arrays
> (a) Type \begin{array}.
> (b) Use an argument to describe how you want your table to be justified. **Immediately following the
\begin{array} command, add a set of brackets.**

So for figure environment specifically I don't know, but in general it could happen - that is what I thought first that `%[{` was about that, but missing the surrounding square brackets.

**Reply by MichelDSchmid** [reply]
Thank you very much for your unbelivable fast help

in quarto I enter now: `![Voltaische Säule](./images/voltaicPillar.jpg){ width="4cm" fig-env="floatingfigure" fig-pos="[r]{6cm}" }`
and in the .tex file I can see: `\begin{floatingfigure}[r]{6cm}`

Everything seems to be ok in the pdf.

Thank you very much
Mike
